sjsonnet/src/sjsonnet/YamlRenderer.scala

```diff
@@ -0,0 +1,554 @@
+      depth -= 1
+      out
+    } else {
+      ujson.Renderer.escape(out, s, unicode = true)
+      out
+    }
+  }
+  override def visitFloat64(d: Double, index: Int) = {
+    flushBuffer()
+    out.append(RenderUtils.renderDouble(d))
+    out
+  }
+  override val colonSnippet = ": "
+  override def flushBuffer() = {
+    if (newlineBuffered) {
+      // drop space between colon and newline
+      if (outBuffer.length() > 1 && outBuffer.charAt(outBuffer.length() - 1) == ' ') {
+        outBuffer.setLength(outBuffer.length() - 1)
+      }
+      PrettyYamlRenderer.writeIndentation(out, indent * depth)
+    }
+    if (dashBuffered) out.append("- ")
+    dashBuffered = false
+    newlineBuffered = false
+    dashBuffered = false
+  }
+  override def visitArray(length: Int, index: Int) = new ArrVisitor[StringWriter, StringWriter] {
+    var empty = true
+    flushBuffer()
+
+    if (!topLevel) {
+      depth += 1
+      newlineBuffered = true
+    }
+    topLevel = false
+
+    val dedentInObject = afterKey && !indentArrayInObject
+    afterKey = false
+    if (dedentInObject) depth -= 1
+    dashBuffered = true
+
+    def subVisitor = YamlRenderer.this
+    def visitValue(v: StringWriter, index: Int): Unit = {
+      empty = false
+      flushBuffer()
+      newlineBuffered = true
+      dashBuffered = true
+    }
+    def visitEnd(index: Int) = {
+      if (!dedentInObject) depth -= 1
+      if (empty) out.append("[]")
+      newlineBuffered = false
+      dashBuffered = false
+      out
+    }
+  }
+  override def visitObject(length: Int, index: Int) = new ObjVisitor[StringWriter, StringWriter] {
+    var empty = true
+    flushBuffer()
+    if (!topLevel) depth += 1
+    topLevel = false
+
+    if (afterKey) newlineBuffered = true
+
+    def subVisitor = YamlRenderer.this
+    def visitKey(index: Int) = YamlRenderer.this
+    def visitKeyValue(s: Any): Unit = {
+      empty = false
+      flushBuffer()
+      out.append(colonSnippet)
+      afterKey = true
+      newlineBuffered = false
+    }
+    def visitValue(v: StringWriter, index: Int): Unit = {
+      newlineBuffered = true
+      afterKey = false
+    }
+    def visitEnd(index: Int) = {
+      if (empty) out.append("{}")
+      newlineBuffered = false
+      depth -= 1
+      flushBuffer()
+      out
+    }
+  }
+}
+
+/**
+ * A version of YamlRenderer that tries its best to make the output YAML as
+ * pretty as possible: unquoted strings, de-dented lists, etc. Follows the PyYAML
+ * style. Also adds the ability to stream writes to a generic `Writer`.
+ */
+class PrettyYamlRenderer(out: Writer = new java.io.StringWriter(),
+                         indentArrayInObject: Boolean = false,
+                         indent: Int = 2,
+                         idealWidth: Int = 80) extends BaseRenderer[Writer](out, indent){
+  var newlineBuffered = false
+  var dashBuffered = false
+  var afterColon = false
+  var afterKey = false
+  var topLevel = true
+  var leftHandPrefixOffset = 0
```

Comment:
This new variable is used to store the width of the LHS of `<key>:` syntax, so that if it's followed by a string that needs wrapping, I can subtract the width of `<key>` to make sure the string wraps at the appropriate place